docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+its supertypes has already been performed.
+
+
+If a method leaves a type off of its signature (parameter type or return type)
+and it does not override or implement anything from a super type, then the
+omitted types are treated as dynamic.
+
+
+Otherwise, the missing types are filled in with the type from the overridden
+method.  If there are multiple overridden/implemented methods and the type to be
+filled in does not agree, it is an error.  If there is no corresponding
+parameter position in the overridden method to infer from, it is treated as
+dynamic (e.g. overriding a one parameter method with a method that takes a
+second optional parameter).
+
+
+### Setter return types
+
+
+If the return type is left off a setter, the return type is inferred to be
+`void`.
+
+
+## Static variable and field inference
+
+
+A “candidate” for inference is any of the following:
+* A top level variable definition
+* A static class level variable definition
+* An instance variable, getter or setter definition with the type annotation
+  omitted.
+
+
+The *inference dependencies of a candidate* which is inferred via override
+inference are the instance fields, setters, and getters that the candidate
+overrides.
+
+
+The *inference dependencies of a candidate* which is inferred via initializer
+inference are the *inference dependencies of the initializer expression* used to
+infer the type of the candidate.
+
+
+The *inference dependencies of an expression*  are as defined below
+as part of initializer inference.
+
+
+Informally, the inference dependencies of a candidate are all of the program
+elements that contribute to inference for the candidate, and hence which must
+have their types inferred before the candidate can have its type inferred.  Note
+that by design, a top level variable or static class variable never depends on
+an instance field, setter, or getter.  This means that top level variable and
+static class variable inference can be done without reference to the results of
+field inference.
+
+
+For each candidate, inference for all of its dependencies must be performed
+before inference for the candidate is performed.  If this cannot be done
+(because of a cycle) it is a static error.  Note that it is always possible for
+the programmer to break a cycle and hence eliminate the error by adding a type
+annotation.
+
+
+### Candidate inference
+
+#### Top level variables
+
+The inferred type of a top level variable is the type inferred from its
+initializer.
+
+#### Static class variables
+
+The inferred type of a static class variable is the type inferred from its
+initializer.
+
+#### Instance field, getter, and setter inference
+
+
+The inferred type of a getter, setter, or field is as follows.  Note that I say
+that a setter overrides a getter if there is a getter of the same name in some
+superclass or interface, and similarly for setters overriding getters, fields,
+etc.
```

Comment:
Might as well use 'we'. Also, it seems natural to mention here that setters/getters induced by instance variable declarations are included: '.. superclass or interface (explicitly declared, or induced by an instance variable declaration), ..'